# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/RTCRtpReceiver.cpp
# Commit: d0b311007c03
# Full Hash: d0b311007c033e83824f5f6996a70ab9e870f31f
# Author: Byron Campen [:bwc] <docfaraday@gmail.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 5
# Description:
#   Bug 1654112 - Get RTCP BYE and RTP timeout handling working again (from Bug 1595479) r=mjf,dminor
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D106145
# ==============================================================================

diff -r daa9eac709dc -r d0b311007c03 dom/media/webrtc/jsapi/RTCRtpReceiver.cpp
--- a/dom/media/webrtc/jsapi/RTCRtpReceiver.cpp	Fri Feb 19 16:20:38 2021 +0000
+++ b/dom/media/webrtc/jsapi/RTCRtpReceiver.cpp	Fri Feb 19 15:56:41 2021 -0600
@@ -101,6 +101,12 @@
   PrincipalHandle principalHandle = GetPrincipalHandle(aWindow, aPrivacyNeeded);
   mTrack = CreateTrack(aWindow, aConduit->type() == MediaSessionConduit::AUDIO,
                        principalHandle.get());
+  // Until Bug 1232234 is fixed, we'll get extra RTCP BYES during renegotiation,
+  // so we'll disable muting on RTCP BYE and timeout for now.
+  if (Preferences::GetBool("media.peerconnection.mute_on_bye_or_timeout",
+                           false)) {
+    aConduit->SetRtcpEventObserver(this);
+  }
   if (aConduit->type() == MediaSessionConduit::AUDIO) {
     mPipeline = new MediaPipelineReceiveAudio(
         mPCHandle, aTransportHandler, mMainThread.get(), mStsThread.get(),
@@ -498,6 +504,7 @@
   ASSERT_ON_THREAD(mMainThread);
   if (mPipeline) {
     mPipeline->Shutdown_m();
+    mPipeline->mConduit->SetRtcpEventObserver(nullptr);
     mPipeline = nullptr;
   }
   mTransceiverImpl = nullptr;
@@ -736,6 +743,10 @@
       aSource, aTimestamp, aRtpTimestamp, aHasLevel, aLevel);
 }
 
+void RTCRtpReceiver::OnRtcpBye() { SetReceiveTrackMuted(true); }
+
+void RTCRtpReceiver::OnRtcpTimeout() { SetReceiveTrackMuted(true); }
+
 void RTCRtpReceiver::SetReceiveTrackMuted(bool aMuted) {
   if (mTrack) {
     // This sets the muted state for mTrack and all its clones.